src/server/server.rs

```diff
@@ -160,168 +194,365 @@ impl Server {
-            // reject if we have a higher max txid
+                    tok: tok,
+                }),
+                _ => None,
+            }.unwrap();
+            vote_res.set_success(true);
+        } else if !self.state.valid_leader() &&
+            vote_req.get_term() >= self.last_learned_term &&
+            ((vote_req.get_last_accepted_txid() >= self.last_accepted_txid &&
+            vote_req.get_last_learned_term() == self.last_learned_term) ||
+            (vote_req.get_last_learned_term() > self.last_learned_term)) {
+            // accept this node as the leader if it has a higher term than
+            // we've ever seen and either one of the following conditions:
+            // 1. it has a higher previous max successful tx term
+            // 2. it has the same previous max successful tx term and at
+            //    least as many entries as we do for it.
+            //
+            // These conditions guarantee that we don't lose acked writes
+            // as long as a majority of our previous nodes stay alive.
+
+            self.highest_term = vote_req.get_term();
+            info!("new leader {}", peer_id);
+            self.state = State::Follower {
+                id: peer_id.clone(),
+                term: vote_req.get_term(),
+                tok: env.tok,
+                leader_addr: env.address.unwrap(),
+                until: time::now().to_timespec().add(*LEADER_DURATION),
+            };
+            info!("{:?}", self.state);
+            vote_res.set_success(true);
+        } else {
+            match self.state.term() {
+                Some(term) =>
+                    vote_res.set_term(term),
+                None => (),
+            }
+
+            vote_res.set_success(false);
+        }
+        res.set_vote_res(vote_res);
+        self.reply(env, ByteBuf::from_slice(
+            &*res.write_to_bytes().unwrap()
+        ));
+    }
+
+    fn handle_append(
+        &mut self,
+        env: Envelope,
+        peer_id: PeerID,
+        append: &Append
+    ) {
+        if self.state.is_leader() {
+            warn!("Leader got append request!  This shouldn't happen.");
+            return;
+        }
+
+        let mut res = PeerMsg::new();
+        res.set_srvid(self.id.clone());
+        let mut append_res = AppendRes::new();
+
+        // verify that we are following this node
+        if self.state.is_following(peer_id.clone()) {
+            // verify that it links
+            if append.get_from_term() == self.last_accepted_term &&
+                append.get_from_txid() == self.last_accepted_txid {
+
+                let mut max_term = self.last_accepted_term;
+                let mut max_txid = self.last_accepted_txid;
+                for vkv in append.get_batch() {
+                    if vkv.get_term() < max_term {
+                        error!("vkv term: {} our max: {}", vkv.get_term(), max_term);
+                        panic!("replication stream has decreasing term");
```

Comment:
We should add a TODO here to note that we should kill the server's process alongside panicking.